Fix for hanging Promise when calling quit() in disconnect

torbjornvatn opened the pull request:
It turns out that the Promise returned by redisClient.quit() neverresolves if the connection to Redis for some reson has been closed. To mitigate this we both try to check whether the connection isstill ready by checking redisConnection.isReady and also add aPromise.race based time out to the quit() function it self.

We have added a separate reconnect.test.ts that attempts to emulate that the connection to the Redis server closed when the disconnect code tries to call the quit() function.

Closes #42

torbjornvatn commented:
Here's an issue in `node-redis` that about the same situation https://github.com/redis/node-redis/issues/2723

It mentions that [QUIT is deprecated](https://redis.io/docs/latest/commands/quit/), but that there is no good alternative yet. There's a safer [close()](https://github.com/redis/node-redis/blob/master/packages/client/lib/client/index.ts#L1074) coming in v5 of the client though.

[This](https://gist.github.com/JonCole/925630df72be1351b21440625ff2671f#file-redis-bestpractices-node-js-md) mentions that adding a `setInterval` triggered `ping()` to keep the connection open, but that seem a bit hacky to me.

kjellmorten commented on src/tests/reconnect.test.ts:
```diff
@@ -0,0 +1,76 @@
+    // Wait for the connection to expire which should trigger an attempt to
+    // disconnect (and client.quit() or client.disconnect() under the hood)
+    // and then reconnect on the next call to transporter.connect()
+    clock.tick(options.connectionTimeout + 1)
```
I wonder if this is really doing anything here, as it is not waiting, just changing the time. I might misunderstand the intention, @torbjornvatn, but won't the `setTimeout()` just below be enough?

kjellmorten commented:
Thanks for a good PR with very thorough explanations, @torbjornvatn!

I think this make sense, although I find it a bit hard to verify. I've published this as a beta release now, as this may make it easier to test. Is it possible for you to try and use version `1.3.0-beta.0` and see if that solves your issue?

Btw, the tests won't run, as the GitHub set up is not using `docker-compose.yml` for spinning up Redis in test. I've changed that, but can't commit to this PR, so I've cloned it and made changes as #44. Here I also changed how you call `PING`, to use a `SERVICE` action, to conform with other Integreat transporters, but the effect is the same.

torbjornvatn commented:
> Thanks for a good PR with very thorough explanations, @torbjornvatn!
> 
> I think this make sense, although I find it a bit hard to verify. I've published this as a beta release now, as this may make it easier to test. Is it possible for you to try and use version `1.3.0-beta.0` and see if that solves your issue?
> 
> Btw, the tests won't run, as the GitHub set up is not using `docker-compose.yml` for spinning up Redis in test. I've changed that, but can't commit to this PR, so I've cloned it and made changes as #44. Here I also changed how you call `PING`, to use a `SERVICE` action, to conform with other Integreat transporters, but the effect is the same.

We can indeed test the beta version. Will report back on how it works.

torbjornvatn commented:
We have tested this for some days in prod now, and it looks like it's working like a dream. 
The quit function timeout kicks in from time to time, but then it reconnects and our app keeps on processing

torbjornvatn commented:
Superseeded by #44
